client/sniffer/src/test/java/org/elasticsearch/client/sniff/SnifferTests.java

```diff
@@ -0,0 +1,547 @@
+        try {
+            //all we need to do is initialize the sniffer, sniffing will start automatically in the background
+            new Sniffer(restClient, hostsSniffer, scheduler, sniffInterval, sniffAfterFailureDelay);
+            assertTrue(latch.await(1000, TimeUnit.MILLISECONDS));
+            assertEquals(iters, futures.size());
+            //the last future is the only one that may not be completed yet, as the count down happens
+            //while scheduling the next round which is still part of the execution of the runnable itself.
+            //we don't take the last item from the list as futures may be ouf of order in there
+            lastFuture.get().get();
+            for (Future<?> future : futures) {
+                assertTrue(future.isDone());
+                future.get();
+            }
+        } finally {
+            executor.shutdown();
+            assertTrue(executor.awaitTermination(1000, TimeUnit.MILLISECONDS));
+        }
+        int totalRuns = hostsSniffer.runs.get();
+        assertEquals(iters, totalRuns);
+        int setHostsRuns = totalRuns - hostsSniffer.failures.get() - hostsSniffer.emptyList.get();
+        verify(restClient, times(setHostsRuns)).setHosts(Matchers.<HttpHost>anyVararg());
+        verifyNoMoreInteractions(restClient);
+    }
+
+    /**
+     * Test that {@link Sniffer#close()} shuts down the underlying {@link Scheduler}, and that such calls are idempotent.
+     * Also verifies that the next scheduled round gets cancelled.
+     */
+    public void testClose() {
+        final Future future = mock(Future.class);
+        long sniffInterval = randomLongBetween(1, Long.MAX_VALUE);
+        long sniffAfterFailureDelay = randomLongBetween(1, Long.MAX_VALUE);
+        RestClient restClient = mock(RestClient.class);
+        final AtomicInteger shutdown = new AtomicInteger(0);
+        Scheduler scheduler = new Scheduler() {
+            @Override
+            public Future<?> schedule(Sniffer.Task task, long delayMillis) {
+                return future;
+            }
+
+            @Override
+            public void shutdown() {
+                shutdown.incrementAndGet();
+            }
+        };
+
+        Sniffer sniffer = new Sniffer(restClient, new MockHostsSniffer(), scheduler, sniffInterval, sniffAfterFailureDelay);
+        assertEquals(0, shutdown.get());
+        int iters = randomIntBetween(3, 10);
+        for (int i = 1; i <= iters; i++) {
+            sniffer.close();
+            verify(future, times(i)).cancel(false);
+            assertEquals(i, shutdown.get());
+        }
+    }
+
+    /**
+     * Test calling {@link Sniffer#sniffOnFailure()} once. The next scheduled sniffing round is cancelled,
+     * a new round is scheduled with delay 0, a following one is scheduled with sniffAfterFailure delay, after
+     * which the ordinary sniffing rounds get scheduled with sniffInterval delay
+     */
+    public void testOnFailureSingleRound() throws Exception {
```

Comment:
onFailure behaviour is really hard to test, too moving parts, so I went for testing different aspects of it with the following three test methods. Suggestions are welcome on whether we think coverage is good enough. Certainly better than before as it was 0% up until now :(